src/engine/sparqlExpressions/StdevExpression.h

```diff
@@ -0,0 +1,163 @@
+// Copyright 2024, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author: Christoph Ullinger <ullingec@informatik.uni-freiburg.de>
+
+#ifndef QLEVER_STDEVEXPRESSION_H
+#define QLEVER_STDEVEXPRESSION_H
+
+#include <cmath>
+#include <functional>
+#include <memory>
+#include <variant>
+
+#include "engine/sparqlExpressions/AggregateExpression.h"
+#include "engine/sparqlExpressions/LiteralExpression.h"
+#include "engine/sparqlExpressions/NaryExpression.h"
+#include "engine/sparqlExpressions/SparqlExpression.h"
+#include "engine/sparqlExpressions/SparqlExpressionTypes.h"
+#include "engine/sparqlExpressions/SparqlExpressionValueGetters.h"
+#include "global/ValueId.h"
+
+namespace sparqlExpression {
+
+namespace detail {
+
+/// The STDEV Expression
+
+// Helper function to extract a double from a NumericValue variant
+auto inline numValToDouble =
+    []<typename T>(T&& value) -> std::optional<double> {
+  if constexpr (ad_utility::isSimilar<T, double> ||
+                ad_utility::isSimilar<T, int64_t>) {
+    return static_cast<double>(value);
+  } else {
+    return std::nullopt;
+  }
+};
+
+// Helper expression: The individual deviation squares. A DeviationExpression
+// over X corresponds to the value (X - AVG(X))^2.
+class DeviationExpression : public SparqlExpression {
+ private:
+  Ptr child_;
+  bool distinct_;
+
+ public:
+  DeviationExpression(bool distinct, Ptr&& child)
+      : child_{std::move(child)}, distinct_{distinct} {}
+
+  // __________________________________________________________________________
+  ExpressionResult evaluate(EvaluationContext* context) const override {
+    auto impl =
+        [this, context](SingleExpressionResult auto&& el) -> ExpressionResult {
+      // Prepare space for result
+      VectorWithMemoryLimit<IdOrLiteralOrIri> exprResult{context->_allocator};
+      std::fill_n(std::back_inserter(exprResult), context->size(),
+                  IdOrLiteralOrIri{Id::makeUndefined()});
+
+      auto devImpl = [&exprResult, context](auto generator) {
+        double sum = 0.0;
+        std::vector<double> childResults = {};
+
+        // Collect values as doubles
+        for (auto& inp : generator) {
+          const auto& s = detail::NumericValueGetter{}(std::move(inp), context);
+          auto v = std::visit(numValToDouble, s);
+          if (v.has_value()) {
+            childResults.push_back(v.value());
+            sum += v.value();
+          }
```

Comment:
This is not quite correct, 
If you have a single nonnumeric element, then your complete standard deviation is undefined. You can't just ignore
the inputs. So you need to do something like tracking a bool with an error value.